swiftide/src/transformers/embed.rs

```diff
@@ -64,28 +67,221 @@ impl BatchableTransformer for Embed {
     ///
     /// If the embedding process fails, the function returns a stream with the error.
     #[tracing::instrument(skip_all, name = "transformers.embed")]
-    async fn batch_transform(&self, nodes: Vec<Node>) -> IndexingStream {
+    async fn batch_transform(&self, mut nodes: Vec<Node>) -> IndexingStream {
         // TODO: We should drop chunks that go over the token limit of the EmbedModel
-        let chunks_to_embed: Vec<String> = nodes.iter().map(Node::as_embeddable).collect();
-
-        self.embed_model
-            .embed(chunks_to_embed)
-            .await
-            .map(|embeddings| {
-                nodes
-                    .into_iter()
-                    // Will panic if the number of embeddings doesn't match the number of nodes
-                    .zip_eq(embeddings)
-                    .map(|(mut n, v)| {
-                        n.vector = Some(v);
-                        n
-                    })
-                    .collect::<Vec<_>>()
-            })
-            .into()
+
+        // EmbeddableTypes grouped by node stored in order of processed nodes.
+        let mut embeddings_keys_groups = VecDeque::with_capacity(nodes.len());
+        // Embeddable data of every node stored in order of processed nodes.
+        let embeddables_data = nodes
+            .iter_mut()
+            .fold(Vec::new(), |mut embeddables_data, node| {
+                let embeddables = node.as_embeddables();
+                let mut embeddables_keys = Vec::with_capacity(embeddables.len());
+                for (embeddable_key, embeddable_data) in embeddables.into_iter() {
+                    embeddables_keys.push(embeddable_key);
+                    embeddables_data.push(embeddable_data);
+                }
+                embeddings_keys_groups.push_back(embeddables_keys);
+                embeddables_data
+            });
+
+        // Embeddings vectors of every node stored in order of processed nodes.
+        let mut embeddings = match self.embed_model.embed(embeddables_data).await {
+            Ok(embeddngs) => VecDeque::from(embeddngs),
+            Err(err) => return IndexingStream::iter(Err(err)),
+        };
+
+        IndexingStream::iter(nodes.into_iter().map(move |mut node| {
+            let Some(embedding_keys) = embeddings_keys_groups.pop_front() else {
+                bail!("Missing embedding data");
+            };
+            let remaining_embeddings = embeddings.split_off(embedding_keys.len());
+            let embedding_values = embeddings.clone();
+            embeddings = remaining_embeddings;
```

Comment:
You're cloning (large) vectors and then reassigning the previous value to what was cloned. Maybe there is an elegant way of solving this with drain. At the very least the clone should be avoidable.

Reply:
Fail. `embeddings` is a deque, so I changed it to just `pop_front()` while iterating over embedding keys.